### PR TITLE
EPMLSTRCMW-131

### PR DESCRIPTION
**Pull request by nikita0levchenko**
Method's and test's for project was added.

### CODE REVIEW

File: portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    with ScalatestRouteTest
+    with PlayJsonSupport
+    with ForAllTestContainer {
+
+  override val container: PostgreSQLContainer = TestContainersUtils.getPostgreSQLContainer()
+  container.start()
+  implicit val config: Config = TestContainersUtils.getConfigForPgContainer(container)
+  private val components: ApplicationComponents = new ApplicationComponents()
+  override protected def beforeAll(): Unit = components.datastorageModule.pipelineDatabaseEngine.updateSchema()
+  val ownerId = "1"
```
Comment (Dmitrii-Shevchenko) [inline portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala]: Make it private

File: portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+  import components.controllerModule.projectController
+  import components.datastorageModule.projectRepository
+  import components.datastorageModule.userRepository
+
+  "ProjectController" when {
+    "getProjectByName" should {
+      "return a project with the same name" in {
+        val dummyUser = TestUserUtils.getDummyUser(ownerId)
+        val dummyProject = TestProjectUtils.getDummyProject("11", ownerId)
```
Comment (Dmitrii-Shevchenko) [inline portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala]: You don't check projectId("11") in your tests, let it generate itself

File: portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+              }
+          )
+      }
+    }
+  }
+
+  "updateProject" should {
+    "return status code NoContend if project was successfully updated" in {
+      val dummyUser = TestUserUtils.getDummyUser(ownerId)
+      val dummyProject = TestProjectUtils.getDummyProject("22",ownerId)
```
Comment (Dmitrii-Shevchenko) [inline portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala]: the same as above

File: portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            status shouldBe StatusCodes.NoContent
+          }
+        })
+    }
+  }
+
+  "deleteProjectById" should {
+    "return project's entity with false value if project was successfully deactivated" in {
+      val dummyUser = TestUserUtils.getDummyUser(ownerId)
+      val dummyProject = TestProjectUtils.getDummyProject("33",ownerId)
```
Comment (Dmitrii-Shevchenko) [inline portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala]: the same as above

File: portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala (hunk 2 above)
Comment (Dmitrii-Shevchenko) [reply portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala]: And every time you create dummyUser, but you need only one in DB

File: portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala (hunk 1 above)
Comment (nikita0levchenko) [reply portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala]: Done

File: portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala (hunk 2 above)
Comment (nikita0levchenko) [reply portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala]: Done

File: portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala (hunk 4 above)
Comment (nikita0levchenko) [reply portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala]: Done

File: portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+  private val ownerId = "1"
+  private val dummyUser = TestUserUtils.getDummyUser(ownerId)
+  import components.controllerModule.projectController
+  import components.datastorageModule.projectRepository
+  import components.datastorageModule.userRepository
+
+  "ProjectController" when {
+    "getProjectByName" should {
+      "return a project with the same name" in {
+        val dummyProject = TestProjectUtils.getDummyProject()
```
Comment (Dmitrii-Shevchenko) [inline portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala]: don't forget to signify owner id for project ```TestProjectUtils.getDummyProject(ownerId = UserId(ownerId))```

File: portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+  private val ownerId = "1"
+  private val dummyUser = TestUserUtils.getDummyUser(ownerId)
+  import components.controllerModule.projectController
+  import components.datastorageModule.projectRepository
+  import components.datastorageModule.userRepository
+
+  "ProjectController" when {
+    "getProjectByName" should {
+      "return a project with the same name" in {
+        val dummyProject = TestProjectUtils.getDummyProject(ProjectId(UUID.randomUUID().toString),UserId(ownerId))
```
Comment (Dmitrii-Shevchenko) [inline portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala]: don't need to send project id explicitly, it can be created by itself

File: portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+                }
+              }
+          )
+      }
+    }
+  }
+
+  "updateProject" should {
+    "return status code NoContend if project was successfully updated" in {
+      val dummyProject = TestProjectUtils.getDummyProject(ProjectId(UUID.randomUUID().toString),UserId(ownerId))
```
Comment (Dmitrii-Shevchenko) [inline portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala]: the same as above

File: portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+            Put("/projects", request) ~> projectController.route(accessToken) ~> check {
+              status shouldBe StatusCodes.NoContent
+            }
+          })
+    }
+  }
+
+  "deleteProjectById" should {
+    "return project's entity with false value if project was successfully deactivated" in {
+      val dummyProject = TestProjectUtils.getDummyProject(ProjectId(UUID.randomUUID().toString),UserId(ownerId))
```
Comment (Dmitrii-Shevchenko) [inline portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala]: the same as above

File: portal/src/it/scala/cromwell/pipeline/controller/UserControllerItTest.scala [hunk 9]
```diff
@@ -37,73 +37,73 @@ class UserControllerItTest
...
-                }
-              }
-          )
-      }
-    }
+//    "getUsersByEmail" should {
+//
+//      "should find newly added user by email pattern" in {
+//        val dummyUser: User = TestUserUtils.getDummyUser()
+//        val userByEmailRequest: String = dummyUser.email
```
Comment (YuliyaSheyko) [inline portal/src/it/scala/cromwell/pipeline/controller/UserControllerItTest.scala]: Why the whole test is commented out? If you're doing it in order to get Travis build it's a not good decision. UserControllerItTest checks if our modules work correctly when they are connected to each other. So without this test it can't be merged.

Comment (nikita0levchenko) [reply portal/src/it/scala/cromwell/pipeline/controller/UserControllerItTest.scala]: Done

File: build.sbt [hunk 10]
```diff
@@ -82,3 +82,13 @@ lazy val controllers =
...
+
+lazy val womtool = (project in file("womtool"))
+  .configs(IntegrationTest)
+  .settings(
+    resolvers += Resolver.bintrayRepo("scalalab", "pipeline"),
+    name := "WomTool",
+    commonSettings,
+    libraryDependencies ++= testDependencies ++ cromwellDependencies,
+    addCommandAlias("testAll", "; test ; it:test")
+  )
```
[the remark this answers is not in the document]
Comment (nikita0levchenko) [reply build.sbt]: I have no idea where did it come from

File: controllers/src/test/scala/cromwell/pipeline/controller/ProjectControllerTest.scala [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+      val request = ProjectUpdateRequest(dummyProject.projectId, dummyProject.name, dummyProject.repository)
+
+      when(projectService.updateProject(request, userId)).thenReturn(Future.successful(1))
+
+      Put("/projects", request) ~> projectController.route(accessToken) ~> check {
+        status shouldBe StatusCodes.NoContent
+      }
+    }
+
+    "return InternalServerError status if projects's id doesn't match" in {
```
[the remark this answers is not in the document]
Comment (nikita0levchenko) [reply controllers/src/test/scala/cromwell/pipeline/controller/ProjectControllerTest.scala]: done

File: repositories/src/main/scala/cromwell/pipeline/datastorage/dao/entry/ProjectEntry.scala [hunk 12]
```diff
@@ -27,9 +27,19 @@ trait ProjectEntry {
     projects.filter(_.projectId === projectId).take(1)
   }
 
+  def getProjectByName = Compiled { name: Rep[String] =>
+    projects.filter(_.name === name).take(1)
+  }
+
   def addProjectAction(project: Project) = projects.returning(projects.map(_.projectId)) += project
```
[the remark this answers is not in the document]
Comment (nikita0levchenko) [reply repositories/src/main/scala/cromwell/pipeline/datastorage/dao/entry/ProjectEntry.scala]: done

File: womtool/src/test/scala/cromwell/pipeline/womtool/WomToolTest.scala [hunk 13]
```diff
@@ -0,0 +1,107 @@
+package cromwell.pipeline.womtool
```
[the remark this answers is not in the document]
Comment (nikita0levchenko) [reply womtool/src/test/scala/cromwell/pipeline/womtool/WomToolTest.scala]: I really don't know how I did this(

File: womtool/src/main/scala/cromwell/pipeline/womtool/WomTool.scala [hunk 14]
```diff
@@ -0,0 +1,64 @@
+package cromwell.pipeline.womtool
```
[the remark this answers is not in the document]
Comment (nikita0levchenko) [reply womtool/src/main/scala/cromwell/pipeline/womtool/WomTool.scala]: I just delete womtool from my project on my branch

File: services/src/test/scala/cromwell/pipeline/service/ProjectServiceTest.scala [hunk 15]
```diff
@@ -1,64 +1,82 @@
...
+        val projectId = ProjectId("projectId")
+        val userId = UserId("userId")
+        val project =
+          Project(
+            projectId = projectId,
+            ownerId = UserId("userId"),
+            name = "projectName",
+            repository = "repositoryName",
+            active = false
+          )
```
[the remark this answers is not in the document]
Comment (nikita0levchenko) [reply services/src/test/scala/cromwell/pipeline/service/ProjectServiceTest.scala]: done

File: portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (nikita0levchenko) [reply portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala]: done

File: services/src/test/scala/cromwell/pipeline/service/ProjectServiceTest.scala [hunk 16]
```diff
@@ -1,64 +1,82 @@
...
+      "return project with corresponding id" in {
+        val projectId = ProjectId("projectId")
+        val project =
+          Project(
+            projectId = projectId,
+            ownerId = UserId("userId"),
+            name = "projectName",
+            repository = "repositoryName",
+            active = false
+          )
```
[the remark this answers is not in the document]
Comment (nikita0levchenko) [reply services/src/test/scala/cromwell/pipeline/service/ProjectServiceTest.scala]: done

File: repositories/src/main/scala/cromwell/pipeline/datastorage/dto/ProjectResponse.scala [hunk 17]
```diff
@@ -0,0 +1,13 @@
...
+
+final case class ProjectResponse(
+  projectId: ProjectId,
+  name: String,
+  active: Boolean
+)
+
+object ProjectResponse {
+  implicit val findProjectResponse: OFormat[ProjectResponse] = Json.format[ProjectResponse]
+}
```
Comment (ilyaonishenko) [inline repositories/src/main/scala/cromwell/pipeline/datastorage/dto/ProjectResponse.scala]: Why it's located in the separate class and not in `Project.scala`?

File: services/src/main/scala/cromwell/pipeline/service/ProjectService.scala [hunk 18]
```diff
@@ -12,22 +13,59 @@ class ProjectService(projectRepository: ProjectRepository)(implicit executionCon
   def getProjectById(projectId: ProjectId): Future[Option[Project]] =
     projectRepository.getProjectById(projectId)
 
-  def addProject(request: ProjectAdditionRequest): Future[ProjectId] = {
+  def getProjectByName(namePattern: String, userId: UserId): Future[Option[Project]] = {
+    val result = projectRepository.getProjectByName(namePattern)
+    result.flatMap {
+      case Some(project) if project.ownerId == userId => result
+      case Some(_)                                    => Future.failed(new ProjectAccessDeniedException)
```
Comment (ilyaonishenko) [inline services/src/main/scala/cromwell/pipeline/service/ProjectService.scala]: Maybe we need to provide more context in this exception...

File: services/src/main/scala/cromwell/pipeline/service/ProjectService.scala [hunk 19]
```diff
@@ -12,22 +13,59 @@ class ProjectService(projectRepository: ProjectRepository)(implicit executionCon
...
+
+  def addProject(request: ProjectAdditionRequest, userId: UserId): Future[ProjectId] = {
     val project =
       Project(
         projectId = ProjectId(UUID.randomUUID().toString),
-        ownerId = request.ownerId,
+        ownerId = userId,
         name = request.name,
-        repository = request.repository,
+        repository = "test_repo",
```
Comment (ilyaonishenko) [inline services/src/main/scala/cromwell/pipeline/service/ProjectService.scala]: `test_repo`?

File: services/src/main/scala/cromwell/pipeline/service/ProjectService.scala [hunk 20]
```diff
@@ -12,22 +13,59 @@ class ProjectService(projectRepository: ProjectRepository)(implicit executionCon
...
+          ) match {
+            case Some(value) => value
+            case None        => Future.failed(new ProjectNotFoundException)
+          }
+      )
 
 }
+
+case class ProjectNotFoundException(message: String = "Project not found") extends RuntimeException(message)
+case class ProjectAccessDeniedException(message: String = "Access denied") extends RuntimeException(message)
```
Comment (ilyaonishenko) [inline services/src/main/scala/cromwell/pipeline/service/ProjectService.scala]: Not the best place for exceptions. Move it to the separate object, please.

File: services/src/test/scala/cromwell/pipeline/service/ProjectServiceTest.scala [hunk 21]
```diff
@@ -1,64 +1,69 @@
...
 
   private val projectRepository = mock[ProjectRepository]
   private val projectService = new ProjectService(projectRepository)
-  private val dummyProject: Project = TestProjectUtils.getDummyProject()
 
   "ProjectServiceTest" when {
 
     "addProject" should {
-      "return id of a new project" taggedAs (Service) in {
+      "return id of a new project" in {
```
Comment (ilyaonishenko) [inline services/src/test/scala/cromwell/pipeline/service/ProjectServiceTest.scala]: Why did you remove tagging?

File: services/src/test/scala/cromwell/pipeline/service/ProjectServiceTest.scala [hunk 22]
```diff
@@ -1,64 +1,69 @@
...
         val request =
           ProjectAdditionRequest(
-            ownerId = dummyProject.ownerId,
-            name = dummyProject.name,
-            repository = dummyProject.repository
+            name = "projectName"
           )
+        val projectId = ProjectId("projectId")
+        val ownerId = UserId("ownerId")
```
Comment (ilyaonishenko) [inline services/src/test/scala/cromwell/pipeline/service/ProjectServiceTest.scala]: Why don't you use `dummyProject`?

File: controllers/src/test/scala/cromwell/pipeline/controller/ProjectControllerTest.scala [hunk 23]
```diff
@@ -0,0 +1,106 @@
...
+import org.scalatestplus.mockito.MockitoSugar
+
+import scala.concurrent.Future
+
+class ProjectControllerTest
+  extends AsyncWordSpec
+    with Matchers
+    with ScalatestRouteTest
+    with MockitoSugar
+    with PlayJsonSupport {
```
[the remark this answers is not in the document]
Comment (nikita0levchenko) [reply controllers/src/test/scala/cromwell/pipeline/controller/ProjectControllerTest.scala]: done

File: portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+  container.start()
+  implicit val config: Config = TestContainersUtils.getConfigForPgContainer(container)
+  private val components: ApplicationComponents = new ApplicationComponents()
+  override protected def beforeAll(): Unit = components.datastorageModule.pipelineDatabaseEngine.updateSchema()
+  private val ownerId = "1"
+  private val dummyUser = TestUserUtils.getDummyUser(ownerId)
+  private def projectId = ProjectId(UUID.randomUUID().toString)
+  import components.controllerModule.projectController
+  import components.datastorageModule.projectRepository
+  import components.datastorageModule.userRepository
```
[the remark this answers is not in the document]
Comment (nikita0levchenko) [reply portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala]: The same import used in UserControllerItTest. I was thinking it's OK.

File: portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala [hunk 25]
```diff
@@ -0,0 +1,86 @@
...
+        val dummyProject = TestProjectUtils.getDummyProject(projectId, UserId(ownerId))
+        userRepository
+          .addUser(dummyUser)
+          .flatMap(
+            _ =>
+              projectRepository.addProject(dummyProject).map { _ =>
+                val accessToken = AccessTokenContent(dummyProject.ownerId.value)
+                Get("/projects?name=" + dummyProject.name) ~> projectController.route(accessToken) ~> check {
+                  status shouldBe StatusCodes.OK
+                  responseAs[Option[Project]] shouldEqual Option(dummyProject)
```
[the remark this answers is not in the document]
Comment (nikita0levchenko) [reply portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala]: done

File: portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala [hunk 26]
```diff
@@ -0,0 +1,86 @@
...
+            val accessToken = AccessTokenContent(dummyUser.userId.value)
+            Put("/projects", request) ~> projectController.route(accessToken) ~> check {
+              status shouldBe StatusCodes.NoContent
+            }
+          })
+    }
+  }
+
+  "deleteProjectById" should {
+    "return project's entity with false value if project was successfully deactivated" in {
```
[the remark this answers is not in the document]
Comment (nikita0levchenko) [reply portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala]: done

File: repositories/src/main/scala/cromwell/pipeline/datastorage/dto/ProjectResponse.scala (hunk 17 above)
Comment (nikita0levchenko) [reply repositories/src/main/scala/cromwell/pipeline/datastorage/dto/ProjectResponse.scala]: Because any way all implicit object's for formatting into json-format will be relocated.

File: portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala (hunk 24 above)
[the remark this answers is not in the document]
Comment (nikita0levchenko) [reply portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala]: And because we using variable components, initialized on 26 line. I need this import, because I define a components variable into this class. I think my action is correct.

File: repositories/src/main/scala/cromwell/pipeline/datastorage/dto/ProjectResponse.scala (hunk 17 above)
Comment (nikita0levchenko) [reply repositories/src/main/scala/cromwell/pipeline/datastorage/dto/ProjectResponse.scala]: Look's like it's useless object. It will be removed.

Comment (nikita0levchenko) [reply repositories/src/main/scala/cromwell/pipeline/datastorage/dto/ProjectResponse.scala]: done

File: services/src/main/scala/cromwell/pipeline/service/ProjectService.scala (hunk 18 above)
Comment (nikita0levchenko) [reply services/src/main/scala/cromwell/pipeline/service/ProjectService.scala]: I added more informative message of this exception

Comment (nikita0levchenko) [reply services/src/main/scala/cromwell/pipeline/service/ProjectService.scala]: done

File: services/src/main/scala/cromwell/pipeline/service/ProjectService.scala (hunk 19 above)
Comment (nikita0levchenko) [reply services/src/main/scala/cromwell/pipeline/service/ProjectService.scala]: It is stub for the project addition operation request. User can create a project only in already existing repository. We discussed it with @mnesvit. May be i understood this incorrectly. Before it repository added from projectAdditionRequest.

File: services/src/main/scala/cromwell/pipeline/service/ProjectService.scala (hunk 20 above)
Comment (nikita0levchenko) [reply services/src/main/scala/cromwell/pipeline/service/ProjectService.scala]: done

File: services/src/test/scala/cromwell/pipeline/service/ProjectServiceTest.scala (hunk 21 above)
Comment (nikita0levchenko) [reply services/src/test/scala/cromwell/pipeline/service/ProjectServiceTest.scala]: It was misss-click. I will return it back.

Comment (nikita0levchenko) [reply services/src/test/scala/cromwell/pipeline/service/ProjectServiceTest.scala]: done

File: services/src/test/scala/cromwell/pipeline/service/ProjectServiceTest.scala (hunk 22 above)
Comment (nikita0levchenko) [reply services/src/test/scala/cromwell/pipeline/service/ProjectServiceTest.scala]: fixed

Comment (nikita0levchenko) [reply services/src/test/scala/cromwell/pipeline/service/ProjectServiceTest.scala]: done

File: build.sbt (hunk 10 above)
[the remark this answers is not in the document]
Comment (nikita0levchenko) [reply build.sbt]: done

File: services/src/main/scala/cromwell/pipeline/service/ProjectService.scala (hunk 19 above)
Comment (nikita0levchenko) [reply services/src/main/scala/cromwell/pipeline/service/ProjectService.scala]: done

**Comment by myazinn** [comment]
Your PR contains a lot of commits that should not be here. Because of it, we are basically unable to review this PR.
@ilyaonishenko already told you a [way to fix this](https://github.com/EpamLifeSciencesTeam/cmwl_pipeline/pull/36/files/0bb88202ef1db561b0ae2dd0e75dc49b0b0ded19#r388631176). Obviously, this has not yet been done :)
